### PR TITLE
hls: new option session reload

### PR DESCRIPTION
**Pull request by back-to**
**TL;DR**

New command to replace the current hls stream with a new one.

`--hls-session-reload-time` New session after a given time.

`--hls-session-reload-segment` New session if a normal playlist fails twice.

`--hls-segment-ignore-number` Allow invalid segment numbers

also reload on **StreamError** 

---

`--hls-session-reload-time`

It will **replace** the current stream with a new **stream**
after the given time, **self.session** will be **updated** at the same time,

The command must **cache** the stream **url** and the stream **name** somewhere in the self.session,
as it's only available in **streamlink_cli**

If someone want's to reuse this command from an API, it must be done there aswell
or it will print a warning message.

**Example**

HLS-URL timesout after 10 minutes
SESSION-RELOAD is set to 9 minutes

The stream will be reloaded/replaced every 9 minutes,
it can never hit the 10 minutes timeout.

---

`--hls-session-reload-segment`

This can be used as a fallback, for wrong set-up time with the other command

and also for errors like this:

```
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
error: Error when reading from stream: Read timeout, exiting
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
[console][info] Stream ended
[console][info] Closing currently open stream...
```

when the stream gets replaces by an invalid stream

normal playlist

```
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-ALLOW-CACHE:YES
#EXT-X-TARGETDURATION:5
#EXT-X-MEDIA-SEQUENCE:41031
#EXTINF:5.000000,
44-41031.ts
#EXTINF:5.000000,
44-41032.ts
#EXTINF:5.000000,
44-41033.ts
```

invalid replacement

```
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-ALLOW-CACHE:NO
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:0
#EXTINF:10.000000,
errors/403/403-000.ts
#EXTINF:10.000000,
errors/403/403-001.ts
#EXTINF:10.000000,
errors/403/403-002.ts
#EXTINF:0.280000,
errors/403/403-003.ts
#EXT-X-ENDLIST
```

---

`--hls-segment-ignore-number`

This allows invalid segment numbers.

required for some streams
if they get different segment numbers on different playlists.

---

allow **seconds passthrough** with **hours_minutes_seconds**

---

**Known issues**

1. plugins with authentication

  It might error because you are already logged in,
  the plugin can't find items that are only presented for guests.

2. MuxedStream

  not implemented, lack of test streams and might not work the same way like this.

3. might not work because of different segment numbers

  use `--hls-segment-ignore-number`

---

I tried to create tests with unittest/requests_mock,

but it does not really work with it.

So for now there are no tests.

---

It's completely build as **opt-in**, so if it's not used it can't error anything.

---

it might need some namechanges, if these are not wanted

---

can be used for this
Fixed https://github.com/streamlink/streamlink/issues/1567

Not directly what was asked but it can be done with this differently
Fixed https://github.com/streamlink/streamlink/issues/664

should allow a reload here
Fixed https://github.com/streamlink/streamlink/issues/1231

ref https://github.com/streamlink/streamlink/issues/1058
ref https://github.com/streamlink/streamlink/issues/1370
ref https://github.com/streamlink/streamlink/issues/1380

### CODE REVIEW

**Comment by sbernsteinHBO** [comment]
As someone who understands Streamlink very well and has a pretty deep understanding of all of the arguments, I find the description of these new arguments to be very confusing/vague - especially "--hls-session-reload-time" -- I don't quite understand what that would do/be useful for?

**Comment by beardypig** [comment]
Definitely a useful feature, I wonder if it is something that we could generalise a bit more. That's something that requires a bit of planning though...

File: src/streamlink_cli/argparser.py [hunk 1]
```diff
@@ -146,6 +146,12 @@ def hours_minutes_seconds(value):
     :param value: hh:mm:ss
     :return: seconds
     """
+    try:
+        if int(value):
```
Comment (beardypig) [inline src/streamlink_cli/argparser.py]: I think you could just do `return int(value)` here, normally it returns an `int` and 0 would be a valid result too.

**Comment by back-to** [comment]
@sbernsteinHBO 
_of course everything could be renamed, to something different_

**_but the feature of this command is_**

that streamlink will **restart/reload** your terminal command

this will
- update your `session`
- overwrite your streaming url

It's usefull for streams that expires after a short time,
they would timeout and streamlink would close them.

But with this command it **won't timeout** it will replace the expired stream before it can timeout,
if you provide a timestamp before the timeout.

---

@beardypig 

> generalise

well everything can be renamed to something else

**Comment by back-to** [comment]
> we could generalise a bit more. That's something that requires a bit of planning though...

**let's do it, like you wanted it to**

streamlink needs some different api endpoints for this though

as this might take longer and it will probably incompatible, **I will close this**.

---

I recreated this feature as a custom plugin with the new Plugin Arguments API,
so I don't really need it in the master branch right now

**Comment by beardypig** [comment]
@back-to OK, the `hours_minutes_seconds` change was worth including though - if you want to push that to master. 

I don't know how to organise it, but I think we need to do some planning for the `Stream` API and how to solve problems like #1048 and #1055.
